### DISCUSSION TITLE
Saving settings

### DISCUSSION THREAD

**Discussion by danilkorotkov** [HomeSpan/HomeSpan#276] [Q&A] [answered]
Somfy example uses code to save calibration data
```
   size_t len;
    if(!nvs_get_blob(somfyNVS,sChannel,NULL,&len)){                       // if found data for this channel in NVS
      nvs_get_blob(somfyNVS,sChannel,&shadeData,&len);                    // retrieve data
    } else {
      nvs_set_blob(somfyNVS,sChannel,&shadeData,sizeof(ShadeData));       // save data
      nvs_commit(somfyNVS);                                               // commit to NVS
    }
```
what if instead of this code just to create any number custom characteristic and after declare it/them with (value, **true**) for autosaving with internal HomeSpan routines?
```
CUSTOMCHAR(…,rollingCode,…)
CUSTOMCHAR(…,raiseTime,…)
CUSTOMCHAR(…,lowerTime,…)
```

**Comment by HomeSpan** [comment, marked as answer]
Yes, I have used this technique in my more recent sketches.  It not only makes saving easy, but if you also [set](https://github.com/HomeSpan/HomeSpan/blob/master/docs/Reference.md#define-custom_charnameuuidpermsformatdefaultvalueminvaluemaxvaluestaticrange) the description (and, optionally, the units and range) of the Characteristic it can be viewed and changed in the Eve app.  For an example, see this [UnitTest](https://github.com/HomeSpan/UnitTest) sketch I developed alongside of version 1.5.0 of HomeSpan in order to facilitate regression-testing  using the same code across the ESP32, ESP32-S2, and ESP32-C3 chips.  There are some subtle differences in the IDF for these chips, as well a very material differences in the peripherals, that HomeSpan needs to take into account.  UnitTest is designed to "stress" various peripherals and different features of HomeSpan so that I can be sure changes I make in a new version don't inadvertently break existing code.

Note that the current version of UnitTest utilizes some features only available in the current dev branch, and so won't be compilable under 1.5.0, but you can still see how Custom Characteristics can be used to accomplish saving/storing.

**Reply by danilkorotkov** [reply]
> it can be viewed and changed in the Eve app.

All of them or only Eve+HomeKit UUIDs?

**Comment by HomeSpan** [comment]
It seems all of the HAP-R2 Services and Characteristics are available in Eve as well.  But Eve does not have undocumented Apple Services, such as the Television Service.  So in the case of UnitTest, I can operate everything from the Eve App if needed.
